### PR TITLE
Outlines

### PR DESCRIPTION
**Pull request by LarsIngo**

### CODE REVIEW

File: src/tracing/marching_squares.cpp [hunk 1]
```diff
@@ -141,23 +152,121 @@ static void marching_squares(bool topLeft, bool topRight, bool bottomRight, bool
     }
 }
 
+/*
+ * Find other vertex with same position.
+ * @param this_index Vertex index of vertex to find other vertex with same position.
+ * @param vertices The vertices.
+ * @param out_found_index Variable to store vertex index of the found vertex.
+ * @todo Optimize using lookup table.
+ */
```
Comment (Chainsawkitten) [inline src/tracing/marching_squares.cpp]: Return value isn't documented.

File: src/tracing/marching_squares.cpp [hunk 2]
```diff
@@ -141,23 +152,121 @@ static void marching_squares(bool topLeft, bool topRight, bool bottomRight, bool
...
+ * @param lines The lines generated from marching squares.
+ * @param out_contour Vector to store vertices.
+ */
+static void trace_contour(const std::size_t vertex_index, std::vector<vertex>& vertices, const std::vector<line>& lines, std::vector<std::size_t>& out_contour) {
+    // Add vertex to contour.
+    out_contour.push_back(vertex_index);
+
+    // Find connected vertex on line.
+    const line& line = lines[vertices[vertex_index].line_index];
+    std::size_t connected_vertex_index = line.vertex_indices[0] == vertex_index ? line.vertex_indices[1] : line.vertex_indices[0];
```
Comment (Chainsawkitten) [inline src/tracing/marching_squares.cpp]: Use parantheses to clarify if this means:
a == (b ? t : f)
or
(a == b) ? t : f

Not strictly necessary but will make code more intuitive, i.m.o.

File: src/tracing/marching_squares.cpp [hunk 3]
```diff
@@ -141,23 +152,121 @@ static void marching_squares(bool topLeft, bool topRight, bool bottomRight, bool
...
+    // Create contours.
+    std::vector<std::vector<std::size_t>> contours;
+    for (std::size_t vertex_index = 0; vertex_index < vertices.size(); ++vertex_index) {
+        if (!vertices[vertex_index].assigned_contour) {
+
+            // Create new contour.
+            contours.push_back(std::vector<std::size_t>());
+            std::vector<std::size_t>& contour = contours.back();
+
+            // Trace contour (assign connected vertex indices to contour).
```
Comment (Chainsawkitten) [inline src/tracing/marching_squares.cpp:204]: Duplicate code that already exists in trace_contour?

**Review by Chainsawkitten — CHANGES_REQUESTED** [review]

File: src/tracing/marching_squares.cpp (hunk 3 above)
Comment (LarsIngo) [reply src/tracing/marching_squares.cpp:204]: Duplicate code was removed

File: src/tracing/marching_squares.cpp [hunk 4]
```diff
@@ -141,23 +152,89 @@ static void marching_kernel(bool topLeft, bool topRight, bool bottomRight, bool
...
+                // Set vertices as assigned to contour.
+                vertices[it_vertex_index].assigned_contour = true;
+                vertices[connected_vertex_index].assigned_contour = true;
+
+                // Find other vertex starting at connected vertex.
+                find_equal_vertex(connected_vertex_index, vertices, it_vertex_index);
+
+            } while (!vertices[it_vertex_index].assigned_contour);
+
+            // Finnish contour by connecting tail and root.
```
Comment (Chainsawkitten) [inline src/tracing/marching_squares.cpp]: We're Swedish.

**Review by Chainsawkitten — CHANGES_REQUESTED** [review]

**Review by Chainsawkitten — APPROVED** [review]

**Pull request merged**